### ISSUE TITLE
How to deal with occlusion and mask exclusion

### ISSUE THREAD

**Issue by pevers** [ultralytics/ultralytics#20503]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi,

I've been working on fine-tuning a YOLOv11 segmentation model to detect and segment metal wires on camera images. The results are really great so far! Here are some examples:

![Image](https://github.com/user-attachments/assets/b9ac7deb-311e-4fdb-a7c8-815b2c009f49)

![Image](https://github.com/user-attachments/assets/61c92cd5-9871-491a-b2dc-8ca484603d85)

## Problem

The question is, how to deal with the occlusion. I've annotated data with "full objects" so that I can distill the centerlines easily even if part of the wire is hidden. I have the layers and the occlusion properties in CVAT correctly annotated. It looks like this:

![Image](https://github.com/user-attachments/assets/60bb9edb-3363-479f-98fa-e2afe51652db)

However, from what I distill in the training data, these properties are not actually used and it is simply a bounding box and a polygon that is fed to the model. So as far as I understand, there is no information which object is in front of another object.

That leads to some cases where the model incorrectly excludes part of the mask. 

![Image](https://github.com/user-attachments/assets/150c27c7-5077-4778-8e38-4db466d19fc7)

When I disable NMS and extract all masks I can only find these:

![Image](https://github.com/user-attachments/assets/1e3a071d-a5d3-4f54-a1ba-305418355361)

![Image](https://github.com/user-attachments/assets/3090f838-d020-43b4-b291-d0cb0c6b3007)

## Training parameters

```
model = YOLO("yolo11x-seg.pt")
results = model.train(
    data="data.yaml",
    epochs=400,
    imgsz=640,
    batch=18,
    project="runs/train",
    name="yolo11x_20250501_copypaste_cutmix",
    workers=18,
    val=True,
    plots=True,
    exist_ok=True,
    copy_paste=0.5,
    cutmix=0.5,
)
```

Dataset is ~10k images (8k training and 2k test)

What can I do to improve the model? **Most** of the time it is correct, but it still frequently happens that it excludes the wrong mask with occlusion. Can I enhance the model to include the depth information?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @pevers, thank you for sharing your detailed results and question with the Ultralytics team! 🚀 It's great to see your success using YOLO11 for segmentation, and the examples you provided are very helpful.

This is an automated response to help get you started. An Ultralytics engineer will review your issue and assist you soon.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) if you haven't already—this helps us debug more efficiently.

For custom training or segmentation ❓ questions, please include as much relevant information as possible, such as:
- Example images from your dataset and annotated cases (which you've already provided—thank you!)
- Training logs or results summaries
- Any configs or scripts used

You may also find helpful tips and best practices for segmentation and custom training in our [Model Training Tips Guide](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth discussions: [Discourse](https://community.ultralytics.com/)
- Community threads: [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Ensure you are running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

You can run YOLO in any of the following verified environments (with dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your contribution! An Ultralytics engineer will follow up with more specific guidance soon 😊

**Comment by Y-T-G** [comment]
You can train with `overlap_masks=False`

**Comment by pevers** [comment]
> You can train with `overlap_mask=False`

Thanks for the suggestion! I think this is what I was looking for! I started a new training with the parameter disabled.

**Comment by glenn-jocher** [comment]
@pevers I'm glad you're trying `overlap_mask=False` - this is exactly what's needed for your wire occlusion scenario. This parameter allows the segmentation masks to properly overlap rather than forcing the model to create non-overlapping masks. 

With occlusions like you have (wires crossing over each other), this setting tells the model it's okay for object masks to overlap in the final prediction, which should help preserve those full wire shapes you've annotated, even when they pass behind other wires.

Let us know how the results look after training with this setting!

**Comment by pevers** [comment]
Thank you @glenn-jocher @Y-T-G , it solves the problem.

![Image](https://github.com/user-attachments/assets/b7b60abf-aa73-4143-912c-8658e5699bef)

**Comment by glenn-jocher** [comment]
Great to hear the solution worked for you, @pevers! Your final result looks excellent - the model is now correctly handling those wire occlusions. This is exactly what `overlap_mask=False` was designed for, allowing objects to maintain their complete shapes even when they overlap. Thanks for sharing your success with the community.
